morph.go

```diff
@@ -209,23 +217,157 @@ func (m *Morph) ApplyDown(limit int) (int, error) {
 
 	var applied int
 	for i := 0; i < steps; i++ {
-		start := time.Now()
 		migrationName := sortedMigrations[i].Name
-		m.config.Logger.Println(formatProgress(fmt.Sprintf(migrationProgressStart, migrationName)))
-
-		down := downMigrations[migrationName]
-		if err := m.driver.Apply(down, true); err != nil {
+		if err := m.apply(downMigrations[migrationName], true); err != nil {
 			return applied, err
 		}
-
 		applied++
-		elapsed := time.Since(start)
-		m.config.Logger.Println(formatProgress(fmt.Sprintf(migrationProgressFinished, migrationName, fmt.Sprintf("%.4fs", elapsed.Seconds()))))
 	}
 
 	return applied, nil
 }
 
+// Diff returns the difference between the applied migrations and the available migrations.
+func (m *Morph) Diff(mode models.Direction) ([]*models.Migration, error) {
+	appliedMigrations, err := m.driver.AppliedMigrations()
+	if err != nil {
+		return nil, err
+	}
+
+	if mode == models.Down {
+		sortedMigrations := reverseSortMigrations(appliedMigrations)
+		downMigrations, err := findDownScripts(sortedMigrations, m.source.Migrations())
+		if err != nil {
+			return nil, err
+		}
+
+		diff := make([]*models.Migration, 0, len(downMigrations))
+		for i := 0; i < len(sortedMigrations); i++ {
+			diff = append(diff, downMigrations[sortedMigrations[i].Name])
+		}
+
+		return diff, nil
+	}
+
+	pendingMigrations, err := computePendingMigrations(appliedMigrations, m.source.Migrations())
+	if err != nil {
+		return nil, err
+	}
+
+	var diff []*models.Migration
+	for _, migration := range sortMigrations(pendingMigrations) {
+		if migration.Direction != models.Up {
+			continue
+		}
```

Comment:
driver.AppliedMigrations always returns Up migrations. https://github.com/mattermost/morph/blob/d876fb543796cded9a08046ff218c80832ecdf0f/drivers/postgres/postgres.go#L289. Same for MySQL as well. Is there any need for this check?

Reply:
Makes sense, I was being a bit cautious.